video/out/vo_sixel.c

```diff
@@ -0,0 +1,483 @@
+    }
+    return SIXEL_OK;
+}
+
+static SIXELSTATUS prepare_dynamic_palette(struct vo *vo)
+{
+    SIXELSTATUS status = SIXEL_FALSE;
+    struct priv *priv = vo->priv;
+
+    /* create histgram and construct color palette
+     * with median cut algorithm. */
+    status = sixel_dither_initialize(priv->testdither, priv->buffer,
+                                     priv->opts->width, priv->opts->height, 3,
+                                     LARGE_NORM, REP_CENTER_BOX,
+                                     QUALITY_LOW);
+    if (SIXEL_FAILED(status))
+        return status;
+
+    if (detect_scene_change(vo)) {
+        if (priv->dither)
+            sixel_dither_unref(priv->dither);
+
+        priv->dither = priv->testdither;
+        status = sixel_dither_new(&priv->testdither, priv->opts->reqcolors, NULL);
+
+        if (SIXEL_FAILED(status))
+            return status;
+        sixel_dither_set_diffusion_type(priv->dither, priv->opts->diffuse);
+    } else {
+        sixel_dither_set_body_only(priv->dither, 1);
+    }
+
+    return status;
+}
+
+static int resize(struct vo *vo)
+{
+    struct priv *priv = vo->priv;
+
+    talloc_free(priv->buffer);
+
+    if (priv->dither) {
+        sixel_dither_unref(priv->dither);
+        priv->dither = NULL;
+    }
+
+    if (priv->testdither) {
+        sixel_dither_unref(priv->testdither);
+        priv->testdither = NULL;
+    }
+    SIXELSTATUS status = sixel_dither_new(&priv->testdither, priv->opts->reqcolors, NULL);
+
+    if (SIXEL_FAILED(status)) {
+        return status;
+    }
+    priv->buffer =
+        talloc_array(NULL, uint8_t, depth * priv->opts->width * priv->opts->height);
+
+    return 0;
+}
+
+static int reconfig(struct vo *vo, struct mp_image_params *params)
+{
+    struct priv *priv = vo->priv;
+    priv->image_height = params->h;
+    priv->image_width  = params->w;
+    priv->image_format = params->imgfmt;
+    scroll_on_demand(vo);
+
+    priv->sws->src = *params;
+    priv->sws->dst = (struct mp_image_params) {
+        .imgfmt = IMGFMT,
+        .w = priv->opts->width,
+        .h = priv->opts->height,
+        .p_w = 1,
+        .p_h = 1,
+    };
+
+    priv->frame = mp_image_alloc(IMGFMT, priv->opts->width, priv->opts->height);
+    if (!priv->frame)
+        return -1;
+
+    if (mp_sws_reinit(priv->sws) < 0)
+        return -1;
+
+    printf(ESC_HIDE_CURSOR);
+    printf(ESC_CLEAR_SCREEN);
+    vo->want_redraw = true;
+
+    return resize(vo);
+}
+
+static void draw_image(struct vo *vo, mp_image_t *mpi)
+{
+    struct priv *priv = vo->priv;
+    struct mp_image src = *mpi;
+    SIXELSTATUS status = SIXEL_FALSE;
+
+    mp_sws_scale(priv->sws, priv->frame, &src);
+
+    memcpy_pic(priv->buffer, priv->frame->planes[0], priv->opts->width * depth, priv->opts->height,
+               priv->opts->width * depth, priv->frame->stride[0]);
+
+    if (priv->opts->fixedpal) {
+        status = prepare_static_palette(vo);
+    } else {
+        status = prepare_dynamic_palette(vo);
```

Comment:
Do I underdstand correctly that by default (dynamic palette) scene change is checked on every frame? how much CPU overhead does it add compared to fixed palette?

Reply:
Yes it is checked for every frame unless `--vo-sixel-fixedpallete=1` is set. I'm not sure how much is the overhead (would be great if you can let me know how to compute that) but I didn't notice significant frame drops.

Reply:
Well, measuring performance is not always easy, but for starters you could check the mpv cpu usage with `htop` or some graphical resource monitor or some such.

Reply:
However, this can also add a meaningful overhead at the terminal itself, because other than detecting the change, when it is detected - it needs to upload a new set of colors into the terminal, and that may or may not be expensive at the terminal side.

How good/bad does it look with a fixed palette and dithering? I'm wondering whether fixed+dither should be the default instead of dynamic. Thoughts?

Reply:
I tried looking at the libsixel functions, but it doesn't seem to be going through image data while computing the dynamic scene change, or while building static palette either. At higher resolutions both methods seemed to have high framedrops. Although in mlterm at least with static palette images aren't distorted. So might be worth switching the default.